src/java.base/windows/native/libjli/java_md.c

```diff
@@ -606,56 +606,70 @@ JLI_ReportErrorMessage(const char* fmt, ...) {
 }
 
 /*
- * Just like JLI_ReportErrorMessage, except that it concatenates the system
- * error message if any, it's up to the calling routine to correctly
- * format the separation of the messages.
+ * Because Windows API errors and regular errno errors can exist at the same time,
+ * compromise and error out when both are present. Windows code should ideally clear
+ * the last error before a call that sets either of these!
  */
 JNIEXPORT void JNICALL
 JLI_ReportErrorMessageSys(const char *fmt, ...)
 {
     va_list vl;
-
-    int save_errno = errno;
     DWORD       errval;
     jboolean freeit = JNI_FALSE;
     char  *errtext = NULL;
 
+    char* const conflict = "Java detected conflicting Windows and C Runtime errors and is unable to provide an accurate report";
```

Comment:
So why not report both of them as part of the message?

Reply:
I was mainly concerned with Thomas's objection that having both would simply be misleading, though now that you say that, maybe this warning combined with both could be a good compromise.

@tstuefe thoughts?

Reply:
Let's back up a bit.

I looked again, and think @dholmes-ora was originally right when he wrote that this coding does not get called on Windows. Not sure why I thought differently, but I cannot find a single callsite on windows.

But that is weird, since originally there was an error reported on Windows: https://bugs.openjdk.org/browse/JDK-8291917 reported a real problem about windows printing useless error messages. What was the original problem again?

If we are right and this code is not needed, I'd just scrap this function on Windows altogether.

Reply:
8291917 was simply about making error messages more informative on Windows when loading the C Runtime or Java Virtual Machine dlls failed (Right now it just prints an unhelpful "Error: loading: ...") - It consequently depended on this RFE for the functionality. Although it _can_ use the other error reporting utilities the JDK has elsewhere, this one happens to be the most convenient, but a more important reason is that the other utilities provided by the JDK all also use the same flawed logic as this one, if we can decide how to fix this function here and now, we'd be able to extrapolate the solution to those other areas as well with very little fuss, so I think fixing this would be better than simply opting to remove it

Reply:
Okay, this is for future use. We don't usually add complexity as a preparation for the future. We usually include usage and functionality in one RFE. That way reviewers can see how the feature is supposed to be used, and whether its worth to keep it.

You may hear other opinions, but what I would do:

- make an RFE that actually enriches failing Win32 API calls with GetLastError, using the non-sys variant (JLI_ReportErrorMessage) and just manually adding the error code at the call sites.
- If that turns out to be many call sites, we can think about adding a windows specific JLI_ReportErrorMessageSysWin32.  
- If that's just 2-5 sites or so, leave it and just scrap JLI_ReportErrorMessageSys for windows.

Just my 5 cent.

Reply:
The (rather incomplete) fix for this was indeed initially part of the original RFE, and only split into another bug afterwards when it was realized there was too much to fit into one PR for an easy review, but I thought it'd be less of a clutter to focus on this one first since the issue already exists in the tracker.

> make an RFE that actually enriches failing Win32 API calls with GetLastError, using the non-sys variant (JLI_ReportErrorMessage) and just manually adding the error code at the call sites.

I don't quite get what you mean?

Reply:
> I don't quite get what you mean?

The only thing JLI_ReportErrorMessageSys has above JLI_ReportErrorMessage is that it prints the error code for you, and that is a very slim win. It turns into a loss quickly if it does that one thing wrong, e.g. by printing unrelated error codes that confuse the analyst. And if that causes us and you to burn cycles with this discussion that could be spent much better elsewhere.

Therefore, what I meant to suggest is to fix whatever you originally wanted to fix with your original JBS issue https://bugs.openjdk.org/browse/JDK-8291917 by calling JLI_ReportErrorMessage directly.

e.g. 
```
p = VirtualAlloc(...) or whatever win32 API you want to cover
if (p == NULL) {
  JLI_ReportErrorMessage("Failed to allocate xxx (%u)", GetLastError());
}
```

Reply:
Initially I did intend to do that, but a raw GetLastError wouldn't exactly be very helpful when a failure does occur, and unfortunately formatting the proper message from it takes quite a bit of ceremony (Thanks for nothing Win32 API!) that would hopefully be confined to only one function - And it seems so far that this one is quite a suitable candidate. To be sure, that RFE doesn't strictly require this one, and can be changed to use other utilities instead, but as mentioned above the same fault is in Windows error reporting routines all over the JDK, and unlike ReportErrorMessageSys which isn't used that often, those are very commonly used across the codebase, by doing that all we'd be doing is sidestepping the problem and leaving it to be fixed later. I did hope that a warning or outright refusal to concat the errors when either GetLastError and errno != 0 was a good solution, though

Reply:
> Initially I did intend to do that, but a raw GetLastError wouldn't exactly be very helpful when a failure does occur

I disagree. Error codes are well documented, unlike posix errno with defined numerical values, and often more helpful than the localized error text.

>, and unfortunately formatting the proper message from it takes quite a bit of ceremony (Thanks for nothing Win32 API!) 

Then factor that out into an utility function similar to strerror and use that as input to JLI_ReportErrorMessage. But again, only if it is worth the complexity. As long as there is not a single use case inside libjli we should not add this function.

>that would hopefully be confined to only one function - And it seems so far that this one is quite a suitable candidate. To be sure, that RFE doesn't strictly require this one, and can be changed to use other utilities instead, but as mentioned above the same fault is in Windows error reporting routines all over the JDK, and unlike ReportErrorMessageSys which isn't used that often, those are very commonly used across the codebase, by doing that all we'd be doing is sidestepping the problem and leaving it to be fixed later.

Yes. And that's fine. Please deal with one problem at a time, libjli in this case, and do a clearly defined solution that fits the problem in scope and complexity. Please follow the advice of reviewers, because you burn valuable reviewer time.

> I did hope that a warning or outright refusal to concat the errors when either GetLastError and errno != 0 was a good solution, though

I disagree. It is more confusing and not worth the added complexity. Look at it that way, at one hand you want to resolve errorcode and think raw errcode values are too much for the lay analyst, on the other hand you want to show him two potentially completely unrelated error texts, together with a cryptic warning, and let him figure out what that means? 

Let's do this simple and right. So far I have not seen a single use of this function inside the libjli codebase. So come up with that first as I wrote before. Then, if you see multiple use cases, we can maybe think of unifying code into a helper function. If there are no usages, just remove the function.